### PR TITLE
fix: issue #5839

### LINKED ISSUE #5839: typescript: empty typeParameters and typeArguments
**Prettier 1.16.4**

Due to inconsistency in reported ast by typescript parser, there is issue with some of typeParameters and typeArguments, in some cases parser outputs `TSTypeParameterInstantiation` and `TSTypeParameterDeclaration` with empty array of `params`

in PR https://github.com/typescript-eslint/typescript-eslint/pull/223i proposed change to make all occurrences of this consistent across all nodes. It means that all occurrences of `<>` are going to lead to crash in this package, as opposed to only sometimes.

`TSTypeParameterInstantiation` and `TSTypeParameterDeclaration` with empty array of `params` should be supported

[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLWcBOAzAhmOAAmwgkOAB0pCbjSAeAPgAoBKAbioF8qQAaEBAAOMVNADOyULkyYIAdwAKMhJJS4AbhFQATfiABGmfAGs4MAMpD86AObIYmAK5wBACxgBbADYB1N6jw4tYEFqqBqBqBAJ7I4OKSAujiWDCKxraeuMh43ikCAFbiAB4AQsZgZpa4nnAAMuhwObh5riBFxRZ23nAAik4Q8M2tAtaYKZhxMNFCcOJgmKgi+kKLsL66MG7IABwADKNyKb7GQnGrc1gaTQKYcACOTqh36biZ2Ui5+SApnqgOzja4m6fQGQ0+LW+MFwBg2Oi2yAATAJHLhUN47ABhCCeLJxKDQG4gJwpAAqMLUXzgXC4QA)
[AST Structure](https://astexplorer.net/#/gist/e4cecffa1a43eb0d6889926c7008df7e/fb9f74643829cec99626c7432a050ca905ff5f36)

**Input:**
```ts
interface foo {
    foo<>();
}
```

**Output:**
```jsx
TypeError: Cannot read property 'pos' of undefined
    at https://prettier.io/lib/parser-typescript.js:1:2122250
    at T (https://prettier.io/lib/parser-typescript.js:1:2123933)
    at t (https://prettier.io/lib/parser-typescript.js:1:2122204)
    at Object.h [as findNextToken] (https://prettier.io/lib/parser-typescript.js:1:2122387)
    at b (https://prettier.io/lib/parser-typescript.js:1:2128676)
    at e (https://prettier.io/lib/parser-typescript.js:1:2155613)
    at d (https://prettier.io/lib/parser-typescript.js:1:2127555)
    at m (https://prettier.io/lib/parser-typescript.js:1:2127698)
    at https://prettier.io/lib/parser-typescript.js:1:2157681
    at Array.map (<anonymous>)
```

**Reference:**
- https://github.com/typescript-eslint/typescript-eslint/issues/220
- https://github.com/typescript-eslint/typescript-eslint/pull/223
- https://github.com/typescript-eslint/typescript-eslint/pull/219

### PR DESCRIPTION
**Pull request by evilebottnawi**
- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fixes #5839

just tests

### CODE REVIEW

File: tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -130,6 +130,56 @@ export const getVehicleDescriptor = async (
...
+                                                                                | printWidth
+=====================================input======================================
+f1<>();
+
+new f2<>();
+
+function f3<>() {}
+
+class f4 {
+    constructor<>() {}
```
Comment (thorn0) [inline tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]: Constructors can't have type parameters, but classes (and interfaces) can.

Comment (evilebottnawi) [reply tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]: @thorn0 can :smile: https://github.com/typescript-eslint/typescript-eslint/pull/223/files#diff-a0bd6d0e4d604ca222ff0af725e5b63fR2

Comment (thorn0) [reply tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]: I think we need a separate folder for tests like this. `typescript_invalid or `typescript_error_recovery`.

File: tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -130,6 +130,56 @@ export const getVehicleDescriptor = async (
...
+}
+
+const f5 = function<>() {}
+
+interface f6 {
```
Comment (thorn0) [inline tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]:
```suggestion
interface f6<> {
```

File: tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -130,6 +130,56 @@ export const getVehicleDescriptor = async (
...
+    test<>();
+}
+
+class f7 {
```
Comment (thorn0) [inline tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]:
```suggestion
class f7<> {
```

File: tests/typescript_generic/empty.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+f1<>();
+
+new f2<>();
+
+function f3<>() {}
+
+class f4 {
```
Comment (thorn0) [inline tests/typescript_generic/empty.ts]:
```suggestion
class f4<> {
```

File: tests/typescript_generic/empty.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    constructor<>() {}
+}
+
+const f5 = function<>() {}
+
+interface f6 {
```
Comment (thorn0) [inline tests/typescript_generic/empty.ts]:
```suggestion
interface f6<> {
```

File: tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (evilebottnawi) [reply tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]: Okey, let's remove that from the PR

Comment (thorn0) [reply tests/typescript_generic/__snapshots__/jsfmt.spec.js.snap]: You misunderstood me. **All** this code is not valid TS. Not only constructors.

**Comment by evilebottnawi** [comment]
@thorn0 ready for review

**Pull request merged**